[SAAS-12272] Show error if selected image is not supported by commcare

ShivamPokhriyal opened the pull request:
Jira: https://dimagi-dev.atlassian.net/browse/SAAS-12272

While analyzing the logs, I found another case of missing attachments on HQ. The issue was happening because the selected image was not supported by commcare. More details in the ticket. 

**Product Note: Improved check for supported multimedia files.**

dimagimon commented:
<!--
  0 Errors
  20 Warnings: <a href="https://github.com/di...
  1 Message: :link: <a href="https://dimagi...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          20 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/activities/FormEntryActivity.java#L635">app/src/org/commcare/activities/FormEntryActivity.java#L635</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/activities/FormEntryActivity.java#L706">app/src/org/commcare/activities/FormEntryActivity.java#L706</a> - Unnecessary; SDK_INT is always &gt;= 16</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/activities/FormEntryActivity.java#L724">app/src/org/commcare/activities/FormEntryActivity.java#L724</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/activities/FormEntryActivity.java#L988">app/src/org/commcare/activities/FormEntryActivity.java#L988</a> - This <code>AsyncTask</code> class should be static or leaks might occur (anonymous org.commcare.tasks.FormLoaderTask)</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/activities/FormEntryActivity.java#L1033">app/src/org/commcare/activities/FormEntryActivity.java#L1033</a> - Unnecessary; SDK_INT is always &gt;= 16</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/activities/FormEntryActivity.java#L1157">app/src/org/commcare/activities/FormEntryActivity.java#L1157</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L112">app/src/org/commcare/utils/FileUtil.java#L112</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L197">app/src/org/commcare/utils/FileUtil.java#L197</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L477">app/src/org/commcare/utils/FileUtil.java#L477</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L507">app/src/org/commcare/utils/FileUtil.java#L507</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L511">app/src/org/commcare/utils/FileUtil.java#L511</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L514">app/src/org/commcare/utils/FileUtil.java#L514</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L602">app/src/org/commcare/utils/FileUtil.java#L602</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L826">app/src/org/commcare/utils/FileUtil.java#L826</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FileUtil.java#L834">app/src/org/commcare/utils/FileUtil.java#L834</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FormUploadUtil.java#L66">app/src/org/commcare/utils/FormUploadUtil.java#L66</a> - <code>Cipher.getInstance</code> should not be called without setting the encryption mode and padding</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FormUploadUtil.java#L138">app/src/org/commcare/utils/FormUploadUtil.java#L138</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FormUploadUtil.java#L243">app/src/org/commcare/utils/FormUploadUtil.java#L243</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FormUploadUtil.java#L244">app/src/org/commcare/utils/FormUploadUtil.java#L244</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FormUploadUtil.java#L301">app/src/org/commcare/utils/FormUploadUtil.java#L301</a> - Using 'Log' instead of 'Timber'</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">:link: <a href="https://dimagi-dev.atlassian.net/browse/SAAS-12272">SAAS-12272</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dimagimon commented on app/src/org/commcare/utils/FormUploadUtil.java:
```diff
@@ -409,7 +409,7 @@ private static String getFileContentType(File f) {
     public static boolean isSupportedMultimediaFile(String filename) {
         for (String ext : SUPPORTED_FILE_EXTS) {
-            if (filename.endsWith(ext)) {
+            if (filename.toLowerCase().endsWith(ext)) {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/943db269bf6832e319807f4f0ed4715c5c38ea7a/app/src/org/commcare/utils/FormUploadUtil.java#L412"></span>Implicitly using the default locale is a common source of bugs: Use <code>toLowerCase(Locale)</code> instead. For strings meant to be internal use <code>Locale.ROOT</code>, otherwise <code>Locale.getDefault()</code>.</td>
    </tr>
  </tbody>
</table>

ShivamPokhriyal replied:
Wow, I don't know about this^. Should I use `toLowerCase(Locale.English)`?

shubham1g5 replied:
yeah we should use `Locale.English` to be safe

damagatchi commented:
[![33% (0.0%) vs master 33%](https://jenkins.dimagi.com/coverage-status-icon/?coverage=0.3268&masterCoverage=0.3268)](https://jenkins.dimagi.com/job/commcare-android-pr-job-builder/462440/)

shubham1g5 commented on app/src/org/commcare/activities/components/ImageCaptureProcessing.java:
```diff
@@ -132,6 +132,12 @@ public static void processImageChooserResponse(FormEntryActivity activity,
             showInvalidImageMessage(activity);
             return;
         }
+        if (!FileUtil.isSupportedMultiMediaFile(activity, selectedImage)) {
```
is the same check implemented for other media capture widgets - audio/video ?

ShivamPokhriyal replied:
Yeah it's present [here](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/activities/FormEntryActivity.java#L406)

shubham1g5 commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -724,13 +724,12 @@ public static void copyFile(InputStream inputStream, File dstFile) throws IOExce
-        } catch (UriToFilePath.NoDataColumnForUriException e) {
-            // No file path available, work with the media uri instead
-            return FormUploadUtil.isSupportedMultimediaFile(media.getPath());
+            String fileName = getFileName(context, media);
```
It feels like `UriToFilePath.getPathFromUri` handles a lot of edge cases around media URIs that we have encountered over the years. Is there a reason we think `getFileName` is better here ? If not, Is it possible to use `getFileName` as an alternative in cases we don't get an extension using `UriToFilePath.getPathFromUri`

ShivamPokhriyal replied:
`getFileName` also uses `UriToFilePath` [here](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/utils/FileUtil.java#L537-L541), but it only executes when the contentResolver couldn't find the display name. 

I'm not sure if `UriToFilePath` gives a better result than contentResolver, but we can improve that and add a case where if the fileName returned by contentResolver doesn't have an extension, we should grab fileName from `UriToFilePath`. Does that sound good?

shubham1g5 replied:
I see, agree with your approach here.

shubham1g5 commented on app/unit-tests/src/org/commcare/utils/FileUtilTest.java:
```diff
@@ -1,32 +0,0 @@
-public class FileUtilTest {
-
-    @Test
-    public void testContentTools() throws UnsupportedEncodingException, Base64DecoderException {
```
why was it removed ?

ShivamPokhriyal replied:
I removed it in this PR #2464 but somehow it got back. I moved this test here https://github.com/dimagi/commcare-android/blob/master/app/unit-tests/src/org/commcare/utils/FileUtilTest.kt#L45-L49 while changing to kotlin file.

shubham1g5 replied:
Sounds good

shubham1g5 reviewed (changes requested):
Hey Shivam, I am not able to find the root cause of the issue in the ticket you linked. I understand that this happens for extensions we don't support though it seems like we already have a [check](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/views/widgets/MediaWidget.java#L217) for those files and these files should not really be getting saved as an answer.

ShivamPokhriyal commented:
>  we already have a [check](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/views/widgets/MediaWidget.java#L217) for those files and these files should not really be getting saved as an answer.

`ImageWidget` doesn't override `MediaWidget`. And, `ImageWidget` doesn't have any [check](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/views/widgets/ImageWidget.java#L314-L323) as well.

ShivamPokhriyal commented:
> Hey Shivam, I am not able to find the root cause of the issue in the ticket you linked.

Yeah, even I don't know the root cause of this. The `MimeTypeMap` should've been able to correctly identify those images, but I'm guessing some manufacturers have a different implementation of `MimeTypeMap` and it doesn't handle those cases. 

So, I've modified the comparison to ignore the character `case`. And also added a check to show a toast message when the image type isn't supported.

shubham1g5 commented:
> ImageWidget doesn't override MediaWidget. And, ImageWidget doesn't have any check as well.

That makes sense. Honestly quite surprised that we don't have this check on `ImageWidget` already since we should have seen this issue much earlier. Do you know if a recent change might have changed the behaviour here ? It would be nice to do a quick test with an old CC version (2.45 probably) to see if this is reproducible there and try to identify the CC version it broke in.

I also think we should do a hotfix for this issue.

ShivamPokhriyal commented:
> That makes sense. Honestly quite surprised that we don't have this check on `ImageWidget` already since we should have seen this issue much earlier. Do you know if a recent change might have changed the behaviour here ? 

I don't think we had a check in `ImageWidget` ever. I think the main reason behind that is anyone coming to the code would think `MediaWidget` to be a superclass of all attachment types. Even I used to believe that at first. 

> It would be nice to do a quick test with an old CC version (2.45 probably) to see if this is reproducible there and try to identify the CC version it might have broke in.

I haven't been able to reproduce the issue on my device. And I think it's only happening on some devices. I think I can grab more data and add it to the jira ticket to analyze how many different users are facing this. 

> I also think we should do a hotfix for this issue.

I was actually planning to add this in a beta release which I'm going to do at the end of this week. I'll add detailed reason in the jira ticket.

shubham1g5 commented:
> I haven't been able to reproduce the issue on my device. And I think it's only happening on some devices.

Ok, Bit confused. Why won't this happen for any file you attach with an unsupported media extension ?

ctsims commented:
I'm fine with hotfixing, although if the bug is affecting a specific set of users I'd also be fine with pushing them onto the pre-release track.

When's our next normal release? It's been a while and might be easier to make go/no-go decisions on hotfixes with the schedule in mind.

damagatchi commented:
[![33% (0.0%) vs master 33%](https://jenkins.dimagi.com/coverage-status-icon/?coverage=0.3276&masterCoverage=0.3276)](https://jenkins.dimagi.com/job/commcare-android-pr-job-builder/462442/)

ShivamPokhriyal commented on app/src/org/commcare/utils/FileUtil.java:
```diff
@@ -544,6 +544,14 @@ public static String getFileName(Context context, Uri uri) throws FileExtensionN
                 cursor.moveToFirst();
                 fileName = cursor.getString(nameIndex);
                 cursor.close();
+                // If file name doesn't have an extension, we should try UriToFilePath
```
I don't like the way I've written this. 
I think we can change this and try to get fileName from `UriToFilePath` first, and then use `ContentResolver` and then `MimeTypeMap`.

shubham1g5 commented:
>  although if the bug is affecting a specific set of users I'd also be fine with pushing them onto the pre-release track.

The reason I think hotifx is the right way to go here is - 

1. Issue leads to a data discrepancy and I would like to have as much users have access to this fix as possible
2. There are around 44 users who have been affected since Jan according to Shivam's log analysis. Reaching out to these many users would be a large support overhead and this list won't be static over time. If these users would have been for a specifc project, this could still have been managed better with a beta release. 

> When's our next normal release?

I was only thinking of planning a release end of this quarter since we haven't had many changes since the last one.

ShivamPokhriyal commented:
@damagatchi retest this please

shubham1g5 reviewed (dismissed):
Nice changes overall :)

ShivamPokhriyal commented:
changed the base branch to `commcare_2.51` for hotfix
